File: java/src/main/java/dev/medzik/libcrypto/X25519.java
package dev.medzik.libcrypto;

import java.security.InvalidKeyException;
import java.security.SecureRandom;

import static org.bouncycastle.math.ec.rfc7748.X25519.POINT_SIZE;
import static org.bouncycastle.math.ec.rfc7748.X25519.calculateAgreement;

public final class X25519 {
    /** Generates a 32-byte private key for Curve25519. */
    public static byte[] generatePrivateKey() {
        byte[] privateKey = new byte[POINT_SIZE];
        org.bouncycastle.math.ec.rfc7748.X25519.generatePrivateKey(new SecureRandom(), privateKey);
        return privateKey;
    }

    /**
     * Computes a 32-byte shared secret between two users.
     * @param ourPrivate 32-byte private key
     * @param theirPublic 32-byte public key
     * @return 32-byte shared secret
     * @throws InvalidKeyException when {@code ourPrivate} is not 32-byte or {@code theirPublic} is invalid.
     */
    public static byte[] computeSharedSecret(byte[] ourPrivate, byte[] theirPublic) throws InvalidKeyException {
        if (ourPrivate.length != POINT_SIZE) {
            throw new InvalidKeyException("Private key must have 32 bytes.");
        }

        if (theirPublic.length != POINT_SIZE) {
            throw new InvalidKeyException("Public key must have 32 bytes.");
        }

        byte[] encoded = new byte[POINT_SIZE];
        calculateAgreement(ourPrivate, 0, theirPublic, 0, encoded, 0);
        return encoded;
    }

    /**
     * Recovers public key from a private key
     * @param privateKey 32-byte private key
     * @return 32-byte public key
     * @throws InvalidKeyException when the {@code privateKey} is not 32 bytes.
     */
    public static byte[] publicFromPrivate(byte[] privateKey) throws InvalidKeyException {
        if (privateKey.length != POINT_SIZE) {
            throw new InvalidKeyException("Private key must have 32 bytes.");
        }

        byte[] publicKey = new byte[POINT_SIZE];
        org.bouncycastle.math.ec.rfc7748.X25519.generatePublicKey(privateKey, 0, publicKey, 0);
        return publicKey;
    }
}

File: javascript/src/x25519.ts
import { Buffer } from "buffer/"
import * as curve25519 from "@stablelib/x25519"

/** Generate a new Curve25519 key pair. */
export function generateKeyPair() {
    const keyPair = curve25519.generateKeyPair();

    return {
        privateKey: bytesToHex(keyPair.secretKey),
        publicKey: bytesToHex(keyPair.publicKey)
    }
}

/**
 * Computes a 32-byte shared secret between two users.
 * @param ourPrivate 32-byte private key
 * @param theirPublic 32-byte public key
 * @return string 32-byte shared secret
 */
export function computeSharedSecret(ourPrivate: string, theirPublic: string) {
    const sharedSecret = curve25519.sharedKey(hexToBuffer(ourPrivate), hexToBuffer(theirPublic));

    return bytesToHex(sharedSecret);
}

/**
 * Recovers public key from a private key
 * @param privateKey 32-byte private key
 * @return string 32-byte public key
 */
export function publicFromPrivate(privateKey: string) {
    const publicKey = curve25519.scalarMultBase(hexToBuffer(privateKey));

    return bytesToHex(publicKey);
}

function hexToBuffer(hex: string) {
    return Buffer.from(hex, 'hex');
}

function bytesToHex(bytes: Uint8Array) {
    return Buffer.from(bytes).toString('hex');
}
